src/client/client.cpp

```diff
@@ -86,57 +94,43 @@ int Client::init() {
     return 0;
 }
 
-// Remember to do error message output for later
-int Client::start(boost::asio::io_context& context) {
-    init();
-
-    // Constrain framerate
-    /* Loop until the user closes the window */
-    while (!glfwWindowShouldClose(window))
-    {
-        processClientInput();
-        processServerInput(context);
-        /* Render here */
-        glClearColor(0.8f, 0.8f, 0.8f, 1.0f);
-        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
-
-        if (this->gameState.phase == GamePhase::GAME) {
-            this->draw();
-        }
+int Client::cleanup() {
+    glDeleteProgram(shaderProgram);
+    return 0;
+}
 
-        /* Swap front and back buffers */
-        glfwSwapBuffers(window);
-        /* Poll for and process events */
-        glfwPollEvents();
+// Handles all rendering
+void Client::displayCallback() {
+    // processInput();
+    /* Render here */
+    glClearColor(0.8f, 0.8f, 0.8f, 1.0f);
+    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
 
-        // std::this_thread::sleep_for(std::chrono::seconds(1));
+    if (this->gameState.phase == GamePhase::GAME) {
+        this->draw();
     }
 
-    glfwTerminate();
-
-    glDeleteProgram(shaderProgram);
-    return 0;
+    /* Poll for and process events */
+    glfwPollEvents();
+    glfwSwapBuffers(window);
 }
 
-void Client::processClientInput() {
-    std::optional<glm::vec3> movement;
-    if(glfwGetKey(window, GLFW_KEY_ESCAPE) == GLFW_PRESS)
-        glfwSetWindowShouldClose(window, true);
-    if(glfwGetKey(window, GLFW_KEY_Q) == GLFW_PRESS)
-        glfwSetWindowShouldClose(window, true);
-    if(glfwGetKey(window, GLFW_KEY_RIGHT) == GLFW_PRESS)
-       movement = glm::vec3(cubeMovementDelta, 0.0f, 0.0f);
-    if(glfwGetKey(window, GLFW_KEY_LEFT) == GLFW_PRESS)
-       movement = glm::vec3(-cubeMovementDelta, 0.0f, 0.0f);
-    if(glfwGetKey(window, GLFW_KEY_UP) == GLFW_PRESS)
-        movement = glm::vec3(0.0f, cubeMovementDelta, 0.0f);
-    if(glfwGetKey(window, GLFW_KEY_DOWN) == GLFW_PRESS)
-        movement = glm::vec3(0.0f, -cubeMovementDelta, 0.0f);
-
-    if (movement.has_value()) {
-        auto eid = 0; 
-        this->session->sendEventAsync(Event(eid, EventType::MoveRelative, MoveRelativeEvent(eid, movement.value())));
-    }
+// Handle any updates 
+void Client::idleCallback(boost::asio::io_context& context) {
+    glm::vec3 movement(0.0f);
+    if(is_held_right)
+        movement += glm::vec3(cubeMovementDelta, 0.0f, 0.0f);
+    if(is_held_left)
+        movement += glm::vec3(-cubeMovementDelta, 0.0f, 0.0f);
+    if(is_held_up)
+        movement += glm::vec3(0.0f, cubeMovementDelta, 0.0f);
+    if(is_held_down)
+        movement += glm::vec3(0.0f, -cubeMovementDelta, 0.0f);
+
+    auto eid = 0; 
+    this->session->sendEventAsync(Event(eid, EventType::MoveRelative, MoveRelativeEvent(eid, movement)));
```

Comment:
Is there a reason the `std::optional` was removed around `movement`? It's not that big of a deal but now it will always be sending this event to the server on every loop iteration (even when there's no movement).

Reply:
That's a good point, initially I changed it because I thought that the flags would usually be in some triggered state but now that you mention it it'll probably help to change it back to the optional.